tests/integration/test_charm.py

```diff
@@ -0,0 +1,276 @@
+        )
+    )
+    await ops_test.model.wait_for_idle(apps=[MYSQL])
+    assert ops_test.model.applications[MYSQL].status == "active"
+    await ops_test.model.add_relation(DATA_INTEGRATOR, MYSQL)
+    await ops_test.model.wait_for_idle(apps=[DATA_INTEGRATOR, MYSQL])
+    assert ops_test.model.applications[DATA_INTEGRATOR].status == "active"
+
+    # get credential for MYSQL
+    credentials = fetch_action_get_credentials(ops_test.model.applications[DATA_INTEGRATOR][0])
+
+    # test connection for MYSQL with retrieved credentials
+    # connection configuration
+    config = {
+        "user": credentials[MYSQL]["username"],
+        "password": credentials[MYSQL]["password"],
+        "host": credentials[MYSQL]["endpoints"].split(":")[0],
+        "database": DATABASE_NAME,
+        "raise_on_warnings": False,
+    }
+
+    with MysqlConnector(config) as cursor:
+
+        create_table_mysql(cursor, DATABASE_NAME)
+
+        insert_data_mysql(
+            cursor,
+            DATABASE_NAME,
+            credentials[MYSQL]["username"],
+            credentials[MYSQL]["password"],
+            credentials[MYSQL]["endpoints"],
+            credentials[MYSQL]["version"],
+            credentials[MYSQL]["read-only-endpoints"],
+        )
+
+        rows = read_data_mysql(cursor, credentials[MYSQL]["username"])
+        first_row = rows[0]
+        # username, password, endpoints, version, ro-endpoints
+        assert first_row[1] == credentials[MYSQL]["username"]
+        assert first_row[2] == credentials[MYSQL]["password"]
+        assert first_row[3] == credentials[MYSQL]["endpoints"]
+        assert first_row[4] == credentials[MYSQL]["version"]
+        assert first_row[5] == credentials[MYSQL]["read-only-endpoints"]
+
+    #  remove relation and test connection again
+    await ops_test.model.applications[DATA_INTEGRATOR].remove_relation(
+        f"{DATA_INTEGRATOR}:mysql", f"{MYSQL}:database"
+    )
+
+    await ops_test.model.wait_for_idle(apps=[MYSQL, DATA_INTEGRATOR])
+    await ops_test.model.add_relation(DATA_INTEGRATOR, MYSQL)
+    await ops_test.model.wait_for_idle(apps=[DATA_INTEGRATOR, MYSQL])
+
+    # join with another relation and check the accessibility of the previously created database
+    new_credentials = await fetch_action_get_credentials(
+        ops_test.model.applications[DATA_INTEGRATOR].units[0]
+    )
+
+    new_config = {
+        "user": new_credentials[MYSQL]["username"],
+        "password": new_credentials[MYSQL]["password"],
+        "host": new_credentials[MYSQL]["endpoints"].split(":")[0],
+        "database": DATABASE_NAME,
+        "raise_on_warnings": False,
+    }
+
+    with MysqlConnector(new_config) as cursor:
+        rows = read_data_mysql(cursor, credentials[MYSQL]["username"])
+        first_row = rows[0]
+        # username, password, endpoints, version, ro-endpoints
+        assert first_row[1] == credentials[MYSQL]["username"]
```

Comment:
given the copy and paste here and above this could be factored out in a single function, even embedded in the test